cli/src/commands/extensions/mod.rs

```diff
@@ -0,0 +1,132 @@
+        )
+        .subcommand(Command::new("list").about("List installed extensions"))
+}
+
+/// Generate the subcommands for each extension.
+/// TODO add tests.
+pub fn extensions_subcommands(command: Command<'_>) -> Command<'_> {
+    let extensions = match installed_extensions() {
+        Ok(extensions) => extensions,
+        Err(e) => {
+            error!("Couldn't list extensions: {}", e);
+            return command;
+        }
+    };
+
+    let names = command
+        .get_subcommands()
+        .map(|n| n.get_name().to_string())
+        .collect::<HashSet<_>>();
+
+    extensions
+        .into_iter()
+        .filter(|ext| {
+            if names.contains(ext.name()) {
+                warn!(
+                    "{}: extension was filtered out due to name conflict",
+                    ext.name()
+                );
+                false
+            } else {
+                true
+            }
+        })
+        .fold(command, |command, ext| {
+            command.subcommand(Command::new(ext.name()))
+        })
+}
+
+/// Entry point for the `extensions` subcommand.
+pub async fn handle_extensions(matches: &ArgMatches) -> CommandResult {
+    if let Some(matches) = matches.subcommand_matches("add") {
+        handle_add_extension(matches.value_of_t("PATH")?).await
+    } else if let Some(matches) = matches.subcommand_matches("remove") {
+        handle_remove_extension(matches.value_of("NAME").unwrap()).await
+    } else {
+        // also covers the `list` case
+        handle_list_extensions().await
+    }
+}
+
+/// Handle the `extension add` subcommand path.
+/// Add the extension from the specified path.
+async fn handle_add_extension(path: PathBuf) -> CommandResult {
+    let extension = Extension::try_from(path)?;
+
+    extension.install()?;
+
+    Ok(CommandValue::Code(ExitCode::Ok))
+}
+
+/// Handle the `extension remove` subcommand path.
+/// Remove the extension named as specified.
+async fn handle_remove_extension(name: &str) -> CommandResult {
+    let extension = Extension::load(name)?;
+
+    extension.uninstall()?;
+
+    Ok(CommandValue::Code(ExitCode::Ok))
+}
+
+/// Handle the `extension` / `extension list` subcommand path.
+/// List installed extensions.
+async fn handle_list_extensions() -> CommandResult {
+    let extensions = installed_extensions()?;
+
+    if extensions.is_empty() {
+        println!("No extensions are currently installed.");
+    } else {
+        extensions.into_iter().for_each(|ext| {
+            println!("{:20}   {}", ext.name(), ext.description().unwrap_or(""));
+        });
+    }
+
+    Ok(CommandValue::Code(ExitCode::Ok))
+}
+
+// Return a list of installed extensions. Filter out invalid extensions instead of exiting early.
+fn installed_extensions() -> Result<Vec<Extension>> {
+    let extensions_path = extensions_path()?;
+
+    if !extensions_path.exists() {
+        return Ok(Vec::new());
+    }
+
+    Ok(std::fs::read_dir(extensions_path)?
```

Comment:
Checking the error returned by `fs::read_dir` would be better than first checking if the directory exists.

```suggestion
    let dir_entry = match fs::read_dir(extensions_path) {
        Ok(d) => d,
        Err(e) => if e.kind() == ErrorKind::NotFound {
            return Ok(Vec::new());
        } else {
            return Err(e.into());
        }
    }

    Ok(dir_entry
```